[FAB-17434] Add `queryapproved` command to query the details of the approved chaincode definition

satota2 opened the pull request:
This patch adds `queryapproved` command for querying the details of the chaincode definition approved by the organization itself to peer CLI, so that Fabric peer can provide a function to query them.
Also, this patch fixes default value of `sequence` parameter to empty for several reasons.

#### Type of change

- New feature
- Bug fix

#### Description

<!--- Describe your changes in detail, including motivation. -->

Currently, Fabric peer does not provide a function to query the details of approved chaincode definitions.

On the other hand, Fabric admins will need to confirm/use the information after approval for their operations.

So, we propose to add a function for querying the details of the approved chaincode definition to Fabric peer CLI (Refer to https://jira.hyperledger.org/browse/FAB-17401).

As a sub-task of FAB-17401, this patch adds `queryapproved` command for querying the details of the chaincode definition approved by the organization itself to peer CLI, so that Fabric peer can provide a function to query them.

Also, this patch fixes default value of `sequence` parameter to `0` that means empty for the following reasons.

- According to `Validate()` in `approveformyorg.go`, this function is implemented assuming that `sequence` is `0`. This is inconsistent with the default value `1`.

- The current default value `1` enables Fabric users `approveformyorg` without specifying the sequence number. Not forcing the specification of a sequence number may cause operation mistakes (e.g, updating the package-id for sequence 1 by mistake) when running `approveformyorg` in the second and subsequent times.

- The default value is preferably `0` to reuse the same parameter `sequence` in CLI commands to query the details of approved chaincode definitions.


The following patches has been merged to extend lifecycle protos and to add `_lifecycle` function for querying approved chaincode:
- https://github.com/hyperledger/fabric-protos/pull/25
- https://github.com/hyperledger/fabric/pull/1220

Immediately after this patch has been merged into the fabric repository,  I will submit the following related patches:
- https://github.com/satota2/fabric/commit/f4d5fff151dc87a597a058184c203b9e9ce72629

#### Related issues
- Main: 
  - https://jira.hyperledger.org/projects/FAB/issues/FAB-17434

- Others: 
  - https://jira.hyperledger.org/projects/FAB/issues/FAB-17401
  - https://jira.hyperledger.org/projects/FAB/issues/FAB-17432
  - https://jira.hyperledger.org/projects/FAB/issues/FAB-17433
  - https://jira.hyperledger.org/projects/FAB/issues/FAB-17435
  - https://jira.hyperledger.org/projects/FAB/issues/FAB-17436

jyellick commented on internal/peer/lifecycle/chaincode/chaincode.go:
```diff
@@ -123,7 +124,7 @@ func ResetFlags() {
 		fmt.Sprint("Time to wait for the event from each peer's deliver filtered service signifying that the 'invoke' transaction has been committed successfully"))
 	flags.StringVarP(&packageID, "package-id", "", "", "The identifier of the chaincode install package")
-	flags.IntVarP(&sequence, "sequence", "", 1, "The sequence number of the chaincode definition for the channel")
+	flags.IntVarP(&sequence, "sequence", "", 0, "The sequence number of the chaincode definition for the channel")
```
This is a change in the default value for this command.  I think I prefer this default, as it works with the defaults provided by the system chaincode (0 is 'unset').  But, @nikhil550 @denyeart will this break anything in our samples/docs etc.?

nikhil550 replied:
None of the samples use the default value for the sequence parameter.

denyeart replied:
I agree with the default change, we'll need to release note it with v2.2.
However I would expect the default change to come in a separate PR, and also the command line help text that mentions the default needs to be updated in the same PR:

`--sequence int                   The sequence number of the chaincode definition for the channel (default 1)`

satota2 replied:
@jyellick @nikhil550 @denyeart
Thank you for your reviews.

Based on the comments, I submitted a separate PR for this as below.
The PR also includes updates to the help docs.
https://github.com/hyperledger/fabric/pull/1262

Immediately after the separate PR has been merged, I will resubmit the updated patch to address the other comments.

jyellick commented on internal/peer/lifecycle/chaincode/queryapproved.go:
```diff
@@ -0,0 +1,215 @@
+func QueryApprovedCmd(a *ApprovedQuerier, cryptoProvider bccsp.BCCSP) *cobra.Command {
+	chaincodeQueryApprovedCmd := &cobra.Command{
+		Use:   "queryapproved",
+		Short: fmt.Sprintf("Query the approved chaincode definition for my org."),
```
This is no longer "for my org", it would be for any org, so perhaps more appropriately "Query an org's approved chaincode definition from its peer." or something like that.

satota2 replied:
@jyellick Thank you for your review.
I will submit the updated patch to address the comments after #1262 is merged.

jyellick reviewed (commented):
This generally looks good to me -- the only thing that gets me a little worried is changing the default of a command flag.  I think it's still probably for the best, but would like to gather some additional opinions.

satota2 committed:
[FAB-17434] Add `queryapproved` command to query the details of the approved chaincode definition

Currently, Fabric peer does not provide a function to query the details of
approved chaincode definitions.

On the other hand, Fabric admins will need to confirm/use the information
after approval for their operations.

This patch adds `queryapproved` command for querying the details of
the chaincode definition approved by the organization itself to peer CLI,
so that Fabric peer can provide a function to query them.

Signed-off-by: Tatsuya Sato <Tatsuya.Sato@hal.hitachi.com>

satota2 commented:
@jyellick
Thank you for merging the PR (#1262).
I updated this patch to address the other comments.
